rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md

```diff
@@ -0,0 +1,394 @@
+    tmp_file_path: PathBuf,
+    stable_file_path: PathBuf,
+    checkpoints: Arc<CheckpointsView>,
+}
+
+pub struct CheckpointsView {
+    checkpoints: DashMap<MultipartUploadKey>,
+}
+
+type MultipartUploadKey = String;
+
+/// Multipart upload state data
+struct MultipartUploadData {
+    /// the identifier for the upload process
+    upload_id: String,
+    /// A list of uploaded parts e_tags. The part number is derived from the index.
+    parts: Vec<String>,
+}
+```
+
+## Drawbacks
+
+- By increasing the time over which events are written to the sink, we
+  are increasing the chances of data loss due to crashes and network
+  interruptions.
+
+## Prior Art
+
+The [S3 output plugin](https://docs.fluentd.org/output/s3) for fluentd
+creates one upload per hour by default. It uses a disk buffer to store
+the object before uploading.
+
+## Alternatives
+
+### Buffer the full large object locally
+
+Instead of using the multipart upload API, Vector could save the data
+that would compose the final object into a local disk buffer, and then
+upload it in the background. This has several disadvantages:
+
+1. It requires local storage for the complete object.
+2. It creates complications for handling upload failures, as
+   acknowledgements for that upload are disconnected from any source
+   data.
+
+### Upload parts as regular objects
+
+Vector could also upload the individual batches as regular objects and
+then go back and assemble the result into a larger object after all
+the uploads are complete. Much of the proposal described above would
+be the same, but the ordering would be changed and an additional step
+is required:
+
+1. Upload batches as regular objects, making note of object names.
+2. When the final object is ready, create it with
+   `CreateMultipartUpload`.
+3. Add the previous objects using `UploadPartCopy`.
+4. Assemble the final object with `CompleteMultipartUpload`.
+5. Delete the intermediate objects that were assembled into the final
+   object.
+
+The downside with this scheme is the additional complexity of tracking
+the intermediate object names and the deletion step at the end of the
+process, both of which are handled automatically when uploading parts
+as above.
+
+### Use the multipart upload copy to effect append behavior
+
+One of the actions possible when creating a multipart object is to
+copy data from an existing object. This can be used to both copy data
+from and replace an object at the same time, allowing for effective
+append (or prepend) behavior without using extra bandwidth. It is
+unclear, however, if this behavior will interact badly with object
+tracking sources like SQS, or if AWS would penalize this action
+pattern for high volume uploads.
+
+## Outstanding Questions
+
+- Should incomplete parts be unconditionally buffered to disk, to
+  allow for producing parts larger than would be feasible in memory,
+  or buffered in memory as usual and only saved to disk to handle
+  persistence across reloads?
```

Comment:
I think so. This also unblocks #11662, since we need to disk buffers to realistically enable e2e ack'ing. Performance would be roughly the same, right?

Reply:
I hadn't thought of the persistent store angle. That's a good point, and it's actually mostly independent of the multipart upload changes.

It would decrease performance slightly, as events would have to be written to disk and then reloaded. For any performance sensitive situation, that data would still be in cache on reload, and may not even have to be flushed to disk by the kernel if we manage it right, but there would still be a small overhead. Since we would be encoding events before dumping to disk, this would not cause any serialization overhead, though.

Ideally, we would also want to change the batch interface to allow for streaming data from a batch, as the current interface assumes everything is purely in-memory. Streaming data asynchronously is possible with hyper, and would be an ideal use case for this. For a 5MB batch, however, it's probably not worth the effort.

Reply:
On the other hand, it's worth asking what is the real difference between an in-sink disk batch buffer and a pre-sink disk buffer? If we aren't going to be uploading parts that are larger than memory (and 5MB is probably not a big concern here), we have an existing mechanism that has at least a similar effect.

Reply:
It does feel like there's a lot of potential conceptual overlap with this and disk buffers. It might be worth laying out explicitly a breakdown of concerns for each. I'm also curious if there's some potential generalization of the existing disk buffer implementation that could accomplish what we're talking about here.